File: 2021/day02/go/part1/part1.go
package part1

import "fmt"
import "strconv"
import "strings"

type Position struct {
	hpos  int
	depth int
}

func ParseLine(line string) (string, int) {
	parts := strings.Split(line, " ")
	dir := parts[0]
	amt, err := strconv.Atoi(parts[1])
	if err != nil {
		panic(err)
	}
	return dir, amt
}

func ComputeAnswer(lines []string) int {
	pos := Position{0, 0}
	for _, line := range lines {
		dir, amt := ParseLine(line)
		switch dir {
		case "forward":
			pos.hpos += amt
		case "down":
			pos.depth += amt
		case "up":
			pos.depth -= amt
		default:
			panic(fmt.Sprintf("Unknown direction: %s", dir))
		}
	}

	if pos.depth < 0 {
		panic(fmt.Sprintf("Invalid depth: %d", pos.depth))
	}

	return pos.hpos * pos.depth
}

File: 2021/day02/rust/src/part1.rs
struct Position {
    hpos: i32,
    depth: i32,
}

pub fn parse_line(line: &str) -> (&str, i32) {
    let tokens: Vec<&str> = line.split(" ").collect();
    let dir = tokens[0];
    let amt = tokens[1].parse().unwrap();
    return (dir, amt);
}

pub fn compute_answer(lines: &Vec<&str>) -> i32 {
    let mut pos = Position { hpos: 0, depth: 0 };
    for index in 0..lines.len() {
        let (dir, amt) = parse_line(lines[index]);
        match dir {
            "forward" => pos.hpos += amt,
            "down" => pos.depth += amt,
            "up" => pos.depth -= amt,
            _ => panic!("Unknown direction: {}", dir),
        }
    }
    if pos.depth < 0 {
        panic!("Invalid depth: {}", pos.depth);
    }
    return pos.hpos * pos.depth;
}
